core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -278,7 +283,7 @@ private CompletionStage<Long> trySizeOptimization(SizeCommand command) {
 
       // Get the size from any shared store that isn't async
       return persistenceManager.size(SHARED.and(NOT_ASYNC))
-            .thenCompose(v -> v >= 0 || !command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)
+            .thenCompose(v -> v >= 0 || (!command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL) && !isLocalCache)
```

Comment:
I suggest to take the second condition out of the lambda to avoid creating it. 

```java
      var stage = persistenceManager.size(SHARED.and(NOT_ASYNC));
      if (!command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL) && !isLocalCache) {
         return stage;
      }
      return stage.thenCompose(v -> v >= 0 ? ...)
```